Implement cancellable coroutines.

radiantgurl opened the pull request:
Implements two hidden methods, `self.cancel_coroutine(function_name: StringName)` and `self.cancel_all_coroutines()`

closes https://github.com/godotengine/godot-proposals/issues/8838

RadiantUwU commented on modules/gdscript/doc_classes/GDScriptFunctionState.xml:
```diff
@@ -0,0 +1,42 @@
+			<return type="Variant" />
+			<param index="0" name="arg" type="Variant" default="null" />
+			<description>
+				Internal method, do not use.
```
To be honest, i dont even see a reason as to why its registered in ClassDB

AThousandShips replied:
Because it's called by the return from `await` and don't think it can be done with other means like callable pointers, it might be possible and in that case this would be more reasonable, but it might not be for legacy reasons

dalexeev commented:
Thanks for the contribution, but let me point out a few things:

As @vnen said on `#gdscript` ([link](https://chat.godotengine.org/channel/gdscript?msg=WY8eyexdJrdaj2jRf)), `GDScriptFunctionState` was made unobtainable in user code **deliberately** ([link 1](https://godotengine.org/article/gdscript-progress-report-new-gdscript-now-merged/#await-instead-of-yield), [link 2](https://github.com/godotengine/godot/pull/39093#issuecomment-636493071)). You cannot get it in a normal way thanks to checks in the analyzer and VM (only through a bug, through interaction with another programming language or similar tricks). The `yield` keyword in 3.x has a number of flaws that were resolved with `await` in 4.x, and obtainable `GDScriptFunctionState` does not fit well with the current `await` design. For example, `await` essentially allows a function to exit before it completes, and the `GDScriptFunctionState` returned in this case may be incompatible with the declared return type. This can be confusing and inconsistent, as can the special behavior of `await` regarding `GDScriptFunctionState` class.

We can't solve this with an `async` modifier for compatibility reasons and because coroutines can be conditional: a `GDScriptFunctionState` is simply not created if `await` has never been executed or its operand was not a signal or a coroutine. Unlike JavaScript, where an `async` function **always** returns a `Promise`.

Also I think a `Promise` class could be implemented on the user side (see godotengine/godot-proposals#5510) or use some alternative approach.

Please note that there are a number of related proposals (see [comment](https://github.com/godotengine/godot-proposals/issues/8838#issuecomment-1881928025)). Thank you for your time, but before implementing any proposal, you should make sure that the problem and solution (idea) are clear to you and other contributors, that there is consensus on our next steps. Coding is exciting and fun, but please try to take the time for planning and discussion if you think a feature might be controversial.

nlupugla commented:
Can you give an example of how this would be used in GDScript?

RadiantUwU commented:
> Can you give an example of how this would be used in GDScript?

Currently at the state of this PR, all you have to do is disable the node and it will automatically cancel all coroutines.

What i have planned? Im planning to add *some* way to be able to cancel coroutines. Similar to how the original godot proposal suggested.

AThousandShips commented:
Why would disabling a node stop coroutines? Code doesn't stop running on disabled nodes, and nothing stops a coroutine from starting on a disabled node does it?

nlupugla commented:
I don't know that coupling process modes and coroutines like that is a good idea. In one of my projects, I have a turn based battle system where I `await` user input. I also use process mode as an easy way to pause the game. If I understand your proposal correctly, it would mean that the user pausing my game would also cancel the coroutine awaiting user input.

AThousandShips commented on modules/gdscript/gdscript.cpp:
```diff
@@ -2008,6 +2010,15 @@ void GDScriptInstance::notification(int p_notification, bool p_reversed) {
+			iterator = iterator->next();
+		}
+	}
+	pending_func_states.clear();
```
This shouldn't be here, it means it clears this when *any* notification is passed

RadiantUwU replied:
Thanks, i think you found the bug.

AThousandShips commented:
Also what happens if you're awaiting the return value of a function? This will break any code that depends on that, potentially crashing valid scripts

RadiantUwU commented:
> Also what happens if you're awaiting the return value of a function? This will break any code that depends on that, potentially crashing valid scripts

Since it emits completed with a null return value when its cancelled, it cannot.

AThousandShips commented:
It might not crash necessarily in some cases, but code will break that expects valid values, so this breaks compatibility in those cases, it's not unsafe or unreasonable to expect the awaited return value to be valid if the method you call works correctly, this changes that expectation

RadiantUwU commented:
> It might not crash necessarily in some cases, but code will break that expects valid values, so this breaks compatibility in those cases, it's not unsafe or unreasonable to expect the awaited return value to be valid if the method you call works correctly, this changes that expectation
> 
> What I'm talking about here is code that already exists or code that doesn't expect this to be a thing, it's unexpected that disabling a node cancels any coroutine

Then would it be fine to simply address directly the feature proposal and implement the said methods directly through GDScriptInstance?

RadiantUwU commented:
Well, currently it doesnt really like the fact it cant be found inside object so GDScript doesnt know how to react to it. How do we find a good implementation for soft methods?

radiantgurl committed:
Implement cancellable coroutines.
Implement two soft methods cancel_all_coroutines and cancel_coroutine

Passed check - disabled from inside
Passed check - disabled inside, multiple coroutines from different threads, same object, different object signals.
